cpp/src/io/utilities/hostdevice_vector.hpp

```diff
@@ -37,71 +40,52 @@ class hostdevice_vector {
  public:
   using value_type = T;
 
-  hostdevice_vector() {}
+  hostdevice_vector() : hostdevice_vector(0, rmm::cuda_stream_default) {}
 
-  hostdevice_vector(hostdevice_vector&& v) { move(std::move(v)); }
-  hostdevice_vector& operator=(hostdevice_vector&& v)
-  {
-    move(std::move(v));
-    return *this;
-  }
-
-  explicit hostdevice_vector(size_t max_size, rmm::cuda_stream_view stream)
-    : hostdevice_vector(max_size, max_size, stream)
+  explicit hostdevice_vector(size_t size, rmm::cuda_stream_view stream)
+    : hostdevice_vector(size, size, stream)
   {
   }
 
   explicit hostdevice_vector(size_t initial_size, size_t max_size, rmm::cuda_stream_view stream)
-    : max_elements(max_size), num_elements(initial_size)
-  {
-    if (max_elements != 0) {
-      CUDF_CUDA_TRY(cudaMallocHost(reinterpret_cast<void**>(&h_data), sizeof(T) * max_elements));
-      d_data.resize(sizeof(T) * max_elements, stream);
-    }
-  }
-
-  ~hostdevice_vector()
+    : d_data(0, stream)
   {
-    if (max_elements != 0) {
-      [[maybe_unused]] auto const free_result = cudaFreeHost(h_data);
-      assert(free_result == cudaSuccess);
-    }
+    CUDF_EXPECTS(initial_size <= max_size, "initial_size can't be larger than max_size");
+    h_data.reserve(max_size);
+    h_data.resize(initial_size);
+    d_data.resize(max_size, stream);
   }
 
-  bool insert(const T& data)
+  void insert(const T& data)
   {
-    if (num_elements < max_elements) {
-      h_data[num_elements] = data;
-      num_elements++;
-      return true;
-    }
-    return false;
+    CUDF_EXPECTS(size() < capacity(), "hostdevice_vector overflow");
```

Comment:
This error can probably be more helpful to the user.
```suggestion
    CUDF_EXPECTS(size() < capacity(), "Cannot insert data into hostdevice_vector because max_size has been exceeded.");
```